data/opam_docs/2.1/04_tricks.md

```diff
@@ -0,0 +1,162 @@
+---
+slug: tricks
+title: Tricks
+description: |
+  Random collection of opam use tricks.
+---
+
+> The following are beyond the scope of the [FAQ](faw), but have been found
+> useful for specific use-cases or for advanced users.
+
+#### Simulate actions from the current switch state (for debugging)
+
+- `opam upgrade --show-actions` (stop at the action summary dialog)
+- `opam upgrade --dry-run` (display only)
+- if you really want to try out the results:
+    * `opam switch export testing-state.export`
+    * `opam switch create tmp-testing --empty`
+    * `opam switch import testing-state.export --fake`
+    * try actions with `--fake` (registers them in opam, but doesn't actually
+      run the build/install commands)
+    * revert to normal: `opam switch <previous>; opam switch remove tmp-testing`
+- Experiment with the solver:
+    * `opam <request> --cudf=cudf-file`
+    * or `opam config cudf-universe >cudf-file-1.cudf`
+    * run e.g. aspcud with `aspcud cudf-file-1.cudf /dev/stdout CRITERIA`
+
+---
+
+#### Install in all switches
+
+Not supported natively at the moment, but it's being considered. Quick hack:
+```
+for switch in $(opam switch list -s); do
+  opam install --switch $switch PACKAGE
+done
+```
+You may want to add `--yes` if you're confident.
+
+---
+
+#### Update opam environment within emacs
+
+The best way is to use `opam-user-setup` (just run `opam user-setup install`),
+and that will make the `opam-update-env` interactive function available in
+emacs.
+
+You could also define the function by hand:
+```lisp
+(defun opam-env ()
+  (interactive nil)
+  (dolist (var (car (read-from-string (shell-command-to-string "opam config env --sexp"))))
+    (setenv (car var) (cadr var))))
+```
+
+You may want to run this at emacs startup if it doesn't inherit the proper shell
+environment. Be also careful that `tuareg-mode` has its own bindings that
+override the environment when using the `compile` command, so that may get in
+the way (see Tuareg [issue #94](https://github.com/ocaml/tuareg/issues/94)).
+
+---
+
+#### Provide a set of packages for a group of users to install
+
+The easiest way is to create a package with your prerequisites as `depends` and
+have them pin that. A quick way to host the file is to use a
+[Gist](https://gist.github.com). Create one with minimal contents and listing
+your packages as dependencies -- the file name **has** to be `opam`:
+
+```
+opam-version: "2.0"
+name: "ocaml101"
+version: "0.1"
+maintainer: "Louis Gesbert <louis.gesbert@ocamlpro.com>"
+depends: [ "menhir" { = "20140422" }
+           "merlin" { >= "2" }
+           "ocp-indent"
+           "ocp-index" ]
+```
+
+Save that and get the `HTTPS clone URL`. All that is needed then is to run:
```

Comment:
```suggestion
Save this to a file and get the `HTTPS clone URL`. Then just run:
```
Not sure this is accurate, so let me know if it's not. I had to read the "Save that" several times before I got what (I think) was meant.